src/main/java/appeng/parts/p2p/CapabilityP2PTunnelPart.java

```diff
@@ -0,0 +1,149 @@
+ * child class!
+ */
+public abstract class CapabilityP2PTunnelPart<P extends CapabilityP2PTunnelPart<P, C>, C> extends P2PTunnelPart<P> {
+    private final Capability<C> capability;
+    // Prevents recursive block updates.
+    private boolean inBlockUpdate = false;
+    // Prevents recursive capability queries.
+    private int capabilityNesting = 0;
+    private final AdjCapability adjCapability = new AdjCapability();
+    private final EmptyAdjCapability emptyAdjCapability = new EmptyAdjCapability();
+    protected C inputHandler;
+    protected C outputHandler;
+    protected C emptyHandler;
+
+    public CapabilityP2PTunnelPart(ItemStack is, Capability<C> capability) {
+        super(is);
+        this.capability = capability;
+    }
+
+    @Override
+    protected float getPowerDrainPerTick() {
+        return 2.0f;
+    }
+
+    public final <T> LazyOptional<T> getCapability(Capability<T> capabilityClass) {
+        if (capabilityClass == capability) {
+            if (isOutput()) {
+                return LazyOptional.of(() -> outputHandler).cast();
+            } else {
+                return LazyOptional.of(() -> inputHandler).cast();
+            }
+        }
+        return LazyOptional.empty();
+    }
+
+    /**
+     * Return the capability connected to side of this P2P connection, or the empty handler if it's not available. The
+     * RAII guard will prevent infinite recursion. Use with try-with-resources!
+     */
+    protected AdjCapability getAdjacentCapability() {
+        capabilityNesting++;
+        return adjCapability;
+    }
+
+    protected class AdjCapability implements AutoCloseable {
+        @Override
+        public void close() {
+            capabilityNesting--;
+        }
+
+        /**
+         * Get the capability, or a null handler if not available. Use within the scope of the enclosing AdjCapability.
+         */
+        protected C get() {
+            if (capabilityNesting == 0) {
+                throw new IllegalStateException("This should be at least 1.");
+            } else if (capabilityNesting == 1) {
+                C adjacentCapability = null;
+                if (isActive()) {
+                    final TileEntity self = getTile();
+                    final TileEntity te = self.getWorld().getTileEntity(self.getPos().offset(getSide().getFacing()));
+
+                    if (te != null) {
+                        adjacentCapability = te.getCapability(capability, getSide().getOpposite().getFacing())
+                                .orElse(null);
+                    }
+                }
+
+                return adjacentCapability == null ? emptyHandler : adjacentCapability;
+            } else {
+                return emptyHandler;
+            }
+        }
+    }
+
+    // Override when there is no capability to retrieve.
+    protected class EmptyAdjCapability extends AdjCapability implements AutoCloseable {
+        @Override
+        public void close() {
```

Comment:
Why does this override close? Isn't the depth already incremented when this object gets constructed hence it leaks the ref?

Reply:
*edit2*: Overall I don't think I like the auto closable use here. In general I'd like it, but it's a bit obtuse with the empty handler in the mix.